File: src/main/com/natpryce/snodge/internal/JsonWalk.java
package com.natpryce.snodge.internal;

import com.google.gson.JsonElement;
import com.natpryce.snodge.JsonPath;

import java.util.Map;
import java.util.stream.Stream;

import static com.natpryce.snodge.internal.JsonFunctions.arrayEntries;
import static java.util.stream.Stream.concat;

public class JsonWalk {
    public static Stream<JsonPath> walk(JsonElement start) {
        return walk(start, JsonPath.root);
    }

    private static Stream<JsonPath> walk(JsonElement element, JsonPath elementPath) {
        return concat(Stream.of(elementPath), walkChildren(element, elementPath));
    }

    private static Stream<JsonPath> walkChildren(JsonElement element, JsonPath elementPath) {
        if (element.isJsonObject()) {
            return walkChildren(elementPath, element.getAsJsonObject().entrySet().stream());
        } else if (element.isJsonArray()) {
            return walkChildren(elementPath, arrayEntries(element.getAsJsonArray()));
        } else {
            return Stream.empty();
        }
    }

    private static <T> Stream<JsonPath> walkChildren(final JsonPath parentPath, Stream<Map.Entry<T, JsonElement>> children) {
        return children.flatMap(child -> walk(child.getValue(), parentPath.extend(child.getKey())));
    }
}

File: src/main/com/natpryce/snodge/json/JsonWalk.kt
@file:JvmName("JsonWalk")

package com.natpryce.snodge.json

import com.natpryce.jsonk.JsonArray
import com.natpryce.jsonk.JsonElement
import com.natpryce.jsonk.JsonObject


fun JsonElement.walk(): Sequence<JsonPath> =
    walk(this, JsonPath.root)

private fun walk(element: JsonElement, elementPath: JsonPath): Sequence<JsonPath> =
    sequenceOf(elementPath) + walkChildren(element, elementPath)

private fun walkChildren(element: JsonElement, elementPath: JsonPath) =
    when (element) {
        is JsonObject ->
            walkChildren(elementPath, objectEntries(element))
        is JsonArray ->
            walkChildren(elementPath, arrayEntries(element))
        else ->
            emptySequence()
    }

private fun <T : Any> walkChildren(parentPath: JsonPath, children: List<Pair<T, JsonElement>>) =
    children.asSequence().flatMap { (key, value) -> walk(value, parentPath.extend(key)) }

private fun objectEntries(element: JsonObject) =
    element.map { it.key to it.value }

private fun arrayEntries(element: JsonArray) =
    element.mapIndexed(::Pair)
